### ISSUE TITLE
SPIRV: Strctured buffers defined in global scope are appearing in descriptor set 0 no matter if you ask for another descriptor set

### ISSUE THREAD

**Issue by Cry-Mory** [Microsoft/DirectXShaderCompiler#1161]
In the hlsl shader code ([Illum@Common_ZPassVS_Common_ZPassVS.hlsl.txt](https://github.com/Microsoft/DirectXShaderCompiler/files/1832723/Illum.Common_ZPassVS_Common_ZPassVS.hlsl.txt)), there is a structured buffer which is defined in the global scope. It is supposed to be in descriptor set = 2, binding = 2 (register(t2, space2)). In the hlsl code, it is defined as follows:

_**StructuredBuffer<SSkinExtraBlendWeights>sb_SkinExtraBlendWeights[1]:register(t2, space2);**_

The _DescriptorSet_ and _Binding_ for sb_SkinExtraBlendWeights do not appear in SPIRV code at all but another descriptor set definition appears on slot 0.1 (DescriptorSet = 0, biding = 1). The SPIRV code can be found [here: Illum@Common_ZPassVS_Common_ZPassVS.spirv.txt](https://github.com/Microsoft/DirectXShaderCompiler/files/1832725/Illum.Common_ZPassVS_Common_ZPassVS.spirv.txt)). The definition for that descriptor set is as follows:

**OpDecorate %_Globals DescriptorSet 0
OpDecorate %_Globals Binding 1**

This causes the program to stuck in GPU driver code for vkCreatePipelineLayout API since the pipeline layout and shader descriptor usages are different. Validation layers also complain about it with the following message:

**_Shader uses descriptor slot 0.1 *used as type `ptr to uniform struct of (arr[1] of struct of (oddtype))` but not declared in pipeline layout_**

**Comment by antiagainst** [comment]
Urgh. This is a regression caused by #1138. Arrays of ConstantBuffers are missed. Will have a fix soon.

**Comment by Cry-Mory** [comment]
Wonderful! Looking forward to it. Thanks a lot 🥇 

I just wanted to point out that the one I mentioned is a structured buffer and not a constant buffer. Maybe they are missed as well.

**Comment by antiagainst** [comment]
Sorry, I posted too quickly. There are actually two issues involved:

1. We are not properly filtering cbuffer/tbuffer when grouping stand-alone global variables into `$Globals`. This is a regression caused by #1138.
2. There are no proper support for arrays of resources (ConstantBuffers/StructuredBuffers/etc.). We have not yet supported this feature.

**Comment by antiagainst** [comment]
#1162 to fix the first issue.

**Comment by antiagainst** [comment]
Actually, I have a question about using arrays of structured buffers. You know that structured buffers are translated into `OpTypeStruct` with one `OpTypeRuntimeArray` member. [`OpTypeRuntimeArray` does not have a known size at compile time.](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#OpTypeRuntimeArray) If having an array of structured buffer, that means we cannot have the suitable `ArrayStride` for the array. But according to [SPIR-V spec](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_validation_rules_for_shader_a_href_capability_capabilities_a), we must have an `ArrayStride`:

> Composite objects in the StorageBuffer, UniformConstant, Uniform, and PushConstant Storage Classes must be explicitly laid out. ...
> * Each array type must have an ArrayStride Decoration.

Correct me if I'm wrong. But I think that effectively forbids using an array of structured buffers in Vulkan at the moment? (ConstantBuffers does not suffer from this because they have no `OpTypeRuntimeArray`s inside.)

But the above is under the assumption of translating the array of structured buffer as a whole and assigning one descriptor to it. If we actually want the compiler to do the magic to split the array and assign a descriptor for each one of them, the problem disappears I think.

**Comment by Cry-Niels** [comment]
AFAIR the array syntax in HLSL describes a range of slots, not an array of descriptors.

Texture2D material[10] : register(t0);

Will use t0 to t9. That's not a Texture-Array, that's an Array of Textures. But maybe I misunderstood.

**Comment by Cry-Mory** [comment]
@Cry-Niels Thanks for your reply.
@antiagainst Thanks for your support and quick response to our request. In our engine shaders, we really don't have an array of structured buffers, texture, ... and it is not really supported by our engine at the moment. So, we can live without it as well.

**Comment by antiagainst** [comment]
@Cry-Niels: Yep, for DX, HLSL `Texture2D[4]` translates into `T0[0:3]`, which is an array of resources that occupy a range of slots. But for Vulkan, a whole array takes only one descriptor. AFAICT, there is no way to assign descriptors to specific array elements. If we want dedicated descriptor for each array element, then we need to generate distinct variables for all array elements. That brings quite some difficulty for the CodeGen since now we need to map access chains on a single variable to multiple ones. And that seems not friendly to `VK_EXT_descriptor_indexing`: https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VK_EXT_descriptor_indexing

**Comment by Cry-Mory** [comment]
@antiagainst So, do you mean that this HLSL feature will not be supported, anytime? Then I think it would be great to document it somewhere.

**Comment by antiagainst** [comment]
@Cry-Mory: It depends.

* For arrays of textures/samplers, yes, we'll support them. Actually we already have some support landed (https://github.com/Microsoft/DirectXShaderCompiler/pull/1184). But there are some details require further touching.
* For arrays of contant/texture buffers, we also have support (https://github.com/Microsoft/DirectXShaderCompiler/pull/1166).
* For arrays of structured buffers, I don't think the Vulkan spec allows declaring arrays of `OpTypeRuntimeArray`s ATM, even with the recently published `VK_EXT_descriptor_indexing` (but my understanding of the spec can certainly be incorrect, please definitely correct me if so); and as described in the above, translating them into distinct variables has problems. So maybe not. The compiler will emit errors if you try to use arrays of structured buffers (enforced by #1166), which I think is a good documentation but we can certainly update SPIR-V.rst. :).
